aws/s3-bucket/bucket.tf

```diff
@@ -0,0 +1,22 @@
+resource "aws_s3_bucket" "bucket" {
+  bucket = var.name
+}
+
+# We want to be able to enable/disable destroy protection from outside this module.
+# Unfortunately lifecycle.prevent_destroy is not supported for modules.
+#
+# To work around this, we create a null_resource that is only created if var.prevent_destroy is set to true.
+# Using a trigger, we have an implicit dependency on the bucket.
+# This way, if var.prevent_destroy is set to true, the bucket can not be destroyed.
+# See https://github.com/hashicorp/terraform/issues/18367 for more information.
+resource "null_resource" "bucket_guardian" {
```

Comment:
QoL: I don't know if I should be happy that you found a workaround or sad that this issue is > 5 years old :D
Anyway, I like the idea to use prevent_destroy!

Thinking out loud, I feel like duplicating code to get a better readability might be helpful here. Using something like (not tested, but you get the point)

```
resource "aws_s3_bucket" "prevent_destroy" {
  count  = var.prevent_destroy ? 1 : 0
  bucket = var.name

  lifecycle {
    prevent_destroy = var.prevent_destroy
  }
}

resource "aws_s3_bucket" "allow_destroy" {
  count  = var.prevent_destroy ? 0 : 1
  bucket = var.name
}


# always use this when referencing to the created s3 bucket from other files.
data "aws_s3_bucket" "main" {
  bucket = var.prevent_destroy ? aws_s3_bucket.prevent_destroy.bucket : aws_s3_bucket.allow_destroy.bucket
}

```

seems (at least to me) to be easier to understand.

Reply:
yes it is easier to understand as long as the resource blocks are quite small. If they become larger, this will lead to a lot of duplicated code. I think introducing this somewhat unintuitive way to set/implement prevent_destroy here now is better in the long run if we start using it within other modules as well.